cohortextractor/query_language.py

```diff
@@ -142,3 +148,17 @@ def __init__(self, source, function, column):
         self.source = source
         self.function = function
         self.column = column
+
+
+def categorise(mapping, default):
+    return ValueFromCategory(mapping, default)
+
+
+CategoryGroup = namedtuple("CategoryGroup", ("source", "operator", "value"))
+
+
+class ValueFromCategory(Value):
+    def __init__(self, groups, default):
+        self.default = default
+        self.source = tuple({group.source for group in groups.values()})
```

Comment:
Do we need `source` as a separate thing here? It feels like it's already captured in `definitions`. We could teach `walk_query_dag` to check the `definitions` dict.

Or maybe another approach would be to add a `get_query_node_references()` method which you can all on any QueryNode instance to get a list of the other query nodes it references. That keeps the logic on the QueryNode subclasses rather than in `walk_query_dag` which is probably the right place for it.

Reply:
I had put it in the query engine at one point and then I took it out to keep it with the `ValuefromCategory`, and as an attribute to make it work with query engine methods that expected a source.  But yes, a query node method would work too and be more explicit